cdc/owner/gc_manager.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package owner
+
+import (
+	"math"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/pkg/config"
+	cdcContext "github.com/pingcap/ticdc/pkg/context"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/tidb/store/tikv/oracle"
+	"go.uber.org/zap"
+)
+
+const (
+	// cdcServiceSafePointID is the ID of CDC service in pd.UpdateServiceGCSafePoint.
+	cdcServiceSafePointID = "ticdc"
+	pdTimeUpdateInterval  = 10 * time.Minute
+)
+
+// gcSafepointUpdateInterval is the minimual interval that CDC can update gc safepoint
+var gcSafepointUpdateInterval = 1 * time.Minute
+
+type gcManager struct {
+	gcTTL int64
+
+	lastUpdatedTime time.Time
+	lastSucceedTime time.Time
+	lastSafePointTs uint64
+
+	pdTimeCached      time.Time
+	lastUpdatedPdTime time.Time
+}
+
+func newGCManager() *gcManager {
+	serverConfig := config.GetGlobalServerConfig()
+	failpoint.Inject("InjectGcSafepointUpdateInterval", func(val failpoint.Value) {
+		gcSafepointUpdateInterval = time.Duration(val.(int) * int(time.Millisecond))
+	})
+	return &gcManager{
+		gcTTL: serverConfig.GcTTL,
+	}
+}
+
+func (m *gcManager) updateGCSafePoint(ctx cdcContext.Context, state *model.GlobalReactorState) error {
+	if time.Since(m.lastUpdatedTime) < gcSafepointUpdateInterval {
+		return nil
+	}
+	minCheckpointTs := uint64(math.MaxUint64)
+	for _, cfState := range state.Changefeeds {
+		if cfState.Info == nil {
+			continue
+		}
+		switch cfState.Info.State {
+		case model.StateNormal, model.StateStopped, model.StateError:
+		default:
+			continue
```

Comment:
Maybe add a debug log here?

Reply:
no need to log this, i think

Reply:
Why do we need a log? @liuzix 
And why not? @leoppro

Reply:
Because there is no need to keep a log here, for no particular reason.

Reply:
My thought was that we could use some log to debug the calculation of gcSafePoint in case of unexpected behavior. But on second thought, the log might produce too many lines and it might not be that useful. So NVM.